Tests on the new implementation: address mypy warnings

MVrachev opened the pull request:
Fixes #1657 
Related to #1582.

**Description of the changes being introduced by the pull request**:

This pr includes:
- configure `mypy` to show error codes
- fixed type annotations in test_metadata_serialization
- two additional invalid serialization tests
- manual fixes for a lot of `mypy` warnings.

When there were `mypy` warnings that we are calling non-annotated function
in annotated context, I decided to add annotations instead of ignoring
those warnings.
That's how I end up adding annotations in the whole tests/utils.py
module.

**NOTE:** Those changes are not done by automatic tool, but instead are manual.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Configure mypy to show error codes

By configuring mypy to show error codes when we get a warning by mypy
we will receive an error code as well.
Those error codes are useful when you want to disable specific mypy
warning for a line with:

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Fix annotations in test_metadata_serialization

In test_metadata_serialization.py "test_case_data" is actually a string
when the decorator calls the actual test functions.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add two more invalid  serialization tests

Those tests are about missing "keys" and "roles" attributes in
Targets.Delegations.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Address mypy warnings

This commit includes manual fixes for a lot of mypy warnings.
When there were warnings that we are calling non-annotated function
in annotated context I decided to add annotations instead of ignoring
those warnings.
That's how I end up adding annotations in the whole tests/utils.py
module.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku reviewed (commented):
Quick comments:
* Can you leave comments explaining the ignores -- they don't seem entirely obvious at first glance?
* Are requests and dateutil really not annotated anywhere ? I wonder how this has not been an issue so far but now requires ignores...
* test_fetcher (and as follow up tuf/requests_fetcher.py) should not need changing as they are part of legacy code

jku commented:
> * Can you leave comments explaining the ignores -- they don't seem entirely obvious at first glance?

I mean PR comments, just for review

MVrachev commented on tests/test_api.py:
```diff
@@ -183,7 +191,7 @@ def test_sign_verify(self):
         targets_key.verify_signature(md_obj, CanonicalJSONSerializer())
         with self.assertRaises(exceptions.UnsignedMetadataError):
-            targets_key.verify_signature(md_obj, JSONSerializer())
+            targets_key.verify_signature(md_obj, JSONSerializer())  # type: ignore[arg-type]
```
Here we receive this warning:
```
tests/test_api.py:194: error: Argument 2 to "verify_signature" of "Key" has incompatible
type "JSONSerializer"; expected "Optional[SignedSerializer]"  [arg-type]
```

I disabled it because the test wanted to test setting an explicit set serializer and which will raise
an `UnsignedMetadataError` and that's what happens.
Please note that `JSONSerializer` inherits the `MetadataSerializer` and not the `SignedSerializer` (as for example the `CanonicalJSONSerializer` does).
If we replace the `JSONSerializer` with the `CanonicalJSONSerializer` the test no longer passes
as we do that same thing on line 192.

jku replied:
isn't this a sign of a bug somewhere then (presumably in the annotations since everything seems to work)?

jku replied:
oh... you mean the test explicitly wants to use a serializer argument that is of a completely different class... but that just happens to have a method of same name `serialize()`.

That's a weird test but not a bug then

MVrachev commented on tests/test_api.py:
```diff
@@ -648,7 +657,7 @@ def test_length_and_hash_validation(self):
             # Test wrong algorithm format (sslib.FormatError)
             snapshot_metafile.hashes = {
-                256: "8f88e2ba48b412c3843e9bb26e1b6f8fc9e98aceb0fbaa97ba37b4c98717d7ab"
+                256: "8f88e2ba48b412c3843e9bb26e1b6f8fc9e98aceb0fbaa97ba37b4c98717d7ab"  # type: ignore[dict-item]
```
Here we receive the following mypy warning:
```
tests/test_api.py:660: error: Dict entry 0 has incompatible type "int": "str";
expected "str": "str"  [dict-item]
```
This error hints we have annotated the `snapshot_metafile.hashes` as a `Dict[str, str]`, but we
are using it as `Dict[int, str]`.
This is expected behavior here as we purposely test an invalid hash algorithm.

MVrachev commented on tuf/requests_fetcher.py:
```diff
@@ -97,7 +98,7 @@ def chunks():
           # time so that the CPU is not hogged in the while loop.
           if settings.SLEEP_BEFORE_ROUND:
-            time.sleep(settings.SLEEP_BEFORE_ROUND)
+            time.sleep(settings.SLEEP_BEFORE_ROUND)  # type: ignore[unreachable]
```
Here I received the error:
```
tuf/requests_fetcher.py:101: error: Statement is unreachable  [unreachable]
```
mypy warns that it's unreachable, because `settings.SLEEP_BEFORE_ROUND` is `None`, but in real
application that could be changed and that's why this warning is unnecessary.

jku replied:
if we wanted to annotate the old code base, the fix for this would be to annotate SLEEP_BEFORE_ROUND correctly... 

but I would rather not do that (and instead remove the tuf/requests_fetcher.py changes from this PR)

MVrachev commented:
> Quick comments:
> 
> * Can you leave comments explaining the ignores -- they don't seem entirely obvious at first glance?

I left comments about each of them.

> * Are requests and dateutil really not annotated anywhere ? I wonder how this has not been an issue so far but now requires ignores...

Hmm, I don't remember why I had to add them to the list in `setup.cfg`. Now when I remove them there aren't any new warnings.
Will remove them.

> * test_fetcher (and as follow up tuf/requests_fetcher.py) should not need changing as they are part of legacy code

I decided to add type annotations for them as I received warnings that I am calling unannotated code in annotated context.
I don't see it as a problem as those changes are small.

jku commented:
> I decided to add type annotations for them as I received warnings that I am calling unannotated code in annotated context.

I would like to avoid touching any more legacy code at least in this PR than is necessary. 

However, I am interested in those warnings -- where do they come from? Why do we end up importing unannotated code that we did not expect to import?

MVrachev commented:
I removed all annotations added to `tuf/request_fetcher.py` and the mypy warnings I received are:
```
tests/test_fetcher.py:61: error: Call to untyped function "RequestsFetcher" in typed context  [no-untyped-call]
tests/test_fetcher.py:75: error: Call to untyped function "fetch" in typed context  [no-untyped-call]
tests/test_fetcher.py:84: error: Call to untyped function "fetch" in typed context  [no-untyped-call]
tests/test_fetcher.py:92: error: Call to untyped function "fetch" in typed context  [no-untyped-call]
tests/test_fetcher.py:101: error: Call to untyped function "fetch" in typed context  [no-untyped-call]
tests/test_fetcher.py:117: error: Call to untyped function "fetch" in typed context  [no-untyped-call]
```
The problem is that we are testing the request_fetcher inside a test for the newer implementation.
Am I correct to consider `test_fetcher.py` as a test file for the new code?

MVrachev commented:
I updated the pr with:
- removing `ignore_missing_imports` from `setup.cfg` for `requests` and `dateutil`
- removed changes for `tuf/request_fetcher.py` and `tests/test_fetcher.py` as they contain code for the legacy code

jku commented on tests/repository_simulator.py:
```diff
@@ -115,9 +111,9 @@ def __init__(self):
 
         # Enable hash-prefixed target file names
-        self.prefix_targets_with_hash = True
+        self.prefix_targets_with_hash: bool = True
```
nitpick: this should not need annotation? or am I missing something?

MVrachev replied:
Yes, you are right. Removing it.

jku commented on tests/test_api.py:
```diff
@@ -47,8 +49,14 @@
     """Tests for public API of all classes in 'tuf/api/metadata.py'."""
 
+    temporary_directory: ClassVar[str]
+    server_process_handler: ClassVar[utils.TestServerProcess]
```
this seems unused?

MVrachev replied:
`temporary_directory` is used inside `tearDownClass`, but about `server_process_handler` you are correct.
Will remove it.

jku commented on tests/test_fetcher_ng.py:
```diff
@@ -136,44 +139,44 @@ def test_response_read_timeout(self, mock_session_get):
 
     # Download a file bigger than expected
-    def test_download_file_length_mismatch(self):
+    def test_download_file_length_mismatch(self) -> Iterator[Any]:
         with self.assertRaises(exceptions.DownloadLengthMismatchError):
             yield self.fetcher.download_file(self.url, self.file_length - 4)
```
this has to be a bug: why would a test ever yield?

(not an important detail though: we expect a raise anyway)

sechkova replied:
It is actually a way to force download_file to execute and raise the error since it is a context manager and returns Iterator[IO]. The other option would be probably:
```
with self.fetcher.download_file(self.url, self.file_length - 4) as f:
    pass
```

MVrachev replied:
I think it's more clear what happens the way it's now.
I added a comment using the explanation @sechkova gave.
Do you think it's better now?

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -115,7 +115,7 @@ def test_targets(self, test_case_data: Tuple[str, bytes, str]):
         # target now exists, is not in cache yet
         info = updater.get_targetinfo(targetpath)
-        self.assertIsNotNone(info)
+        assert(isinstance(info, TargetFile))
```
nitpick: I believe these type of asserts could still just check for none (`assert info is not None`) just in a form that mypy recognises.

sechkova replied:
and why not `self.assertIsInstance` ?

MVrachev replied:
I tried  `self.assertIsInstance` and `mypy` doesn't recognize as it doesn't run our code, but just statically checks it. However, what worked was what @jku suggested using `assert info is not None`.
Will change all places where I use `assert` with this approach.

jku commented on tests/utils.py:
```diff
@@ -48,9 +48,9 @@
-    def real_decorator(function: Callable[[unittest.TestCase, Any], None]):
-        def wrapper(test_cls: unittest.TestCase):
+def run_sub_tests_with_dataset(dataset: DataSet) -> Callable[[Callable], Callable]:
+    def real_decorator(function: Callable[[unittest.TestCase, Any], None]) -> Callable[[unittest.TestCase], None]:
```
this line might fail black once you enable it...

jku commented on tests/utils.py:
```diff
@@ -213,12 +219,12 @@ def _start_process(self, extra_cmd_args, popen_cwd):
     # Run log_queue_worker() in a thread.
     # The thread will exit when the child process dies.
-    self._log_queue = queue.Queue()
+    self._log_queue: queue.Queue = queue.Queue()
```
unlikely to need annotation?

MVrachev replied:
I received a warning before about that, but I can't reproduce it now.
So, yea removing it.

jku reviewed (approved):
Wow, the lack of implicit None return value really makes this a big patch...

Good work: I left some comments but I don't think any are really required fixes. Take a look at the comments but I'm fine with merging as is.
